File: lib/data/source/sensors/garmin/garmin_byte_writer.dart
import 'dart:typed_data';

/// Little-endian write cursor.
///
/// Port of Gadgetbridge's `MessageWriter`. Grows as needed; [toBytes] returns
/// exactly what was written.
class GarminByteWriter {
  GarminByteWriter([int initialCapacity = 256])
      : _view = ByteData(initialCapacity);

  ByteData _view;
  int _pos = 0;

  int get length => _pos;

  void _ensure(int extra) {
    if (_pos + extra <= _view.lengthInBytes) return;
    var next = _view.lengthInBytes * 2;
    while (next < _pos + extra) {
      next *= 2;
    }
    final grown = ByteData(next)
      ..buffer.asUint8List().setRange(0, _pos, _view.buffer.asUint8List());
    _view = grown;
  }

  void writeByte(int value) {
    _ensure(1);
    _view.setUint8(_pos, value & 0xFF);
    _pos += 1;
  }

  void writeShort(int value) {
    _ensure(2);
    _view.setUint16(_pos, value & 0xFFFF, Endian.little);
    _pos += 2;
  }

  void writeInt(int value) {
    _ensure(4);
    _view.setUint32(_pos, value & 0xFFFFFFFF, Endian.little);
    _pos += 4;
  }

  void writeLong(int value) {
    _ensure(8);
    _view.setUint64(_pos, value, Endian.little);
    _pos += 8;
  }

  void writeBytes(Uint8List bytes) {
    _ensure(bytes.length);
    _view.buffer.asUint8List().setRange(_pos, _pos + bytes.length, bytes);
    _pos += bytes.length;
  }

  /// Overwrites 2 bytes at [offset] — used to backfill the frame length once the
  /// payload is written.
  void patchShort(int offset, int value) =>
      _view.setUint16(offset, value & 0xFFFF, Endian.little);

  Uint8List toBytes() => Uint8List.sublistView(_view, 0, _pos);
}

File: app/src/main/kotlin/tech/mmarca/openvitals/devices/garmin/GarminByteWriter.kt
package tech.mmarca.openvitals.devices.garmin

/** Little-endian write cursor. Grows as needed; writers return `this` for chaining. */
class GarminByteWriter(initialCapacity: Int = 256) {

    private var buffer = ByteArray(if (initialCapacity > 0) initialCapacity else 1)

    var length: Int = 0
        private set

    private fun ensure(extra: Int) {
        if (length + extra <= buffer.size) return
        var next = buffer.size * 2
        while (next < length + extra) {
            next *= 2
        }
        buffer = buffer.copyOf(next)
    }

    fun writeByte(value: Int): GarminByteWriter {
        ensure(1)
        buffer[length++] = value.toByte()
        return this
    }

    fun writeShort(value: Int): GarminByteWriter {
        ensure(2)
        buffer[length++] = value.toByte()
        buffer[length++] = (value shr 8).toByte()
        return this
    }

    fun writeInt(value: Long): GarminByteWriter {
        ensure(4)
        for (i in 0 until 4) {
            buffer[length++] = (value shr (8 * i)).toByte()
        }
        return this
    }

    fun writeInt(value: Int): GarminByteWriter = writeInt(value.toLong())

    fun writeLong(value: Long): GarminByteWriter {
        ensure(8)
        for (i in 0 until 8) {
            buffer[length++] = (value shr (8 * i)).toByte()
        }
        return this
    }

    fun writeBytes(bytes: ByteArray): GarminByteWriter {
        ensure(bytes.size)
        bytes.copyInto(buffer, length)
        length += bytes.size
        return this
    }

    /**
     * A length-prefixed UTF-8 string, NUL-terminated inside the length.
     * Truncated to fit the length byte, so a long name cannot corrupt the frame.
     */
    fun writeString(value: String): GarminByteWriter {
        val encoded = value.toByteArray(Charsets.UTF_8)
        val bytes = if (encoded.size > 254) encoded.copyOf(254) else encoded
        writeByte(bytes.size + 1)
        writeBytes(bytes)
        writeByte(0)
        return this
    }

    /** Overwrites 2 bytes at [offset], to backfill the frame length. */
    fun patchShort(offset: Int, value: Int) {
        buffer[offset] = value.toByte()
        buffer[offset + 1] = (value shr 8).toByte()
    }

    fun toBytes(): ByteArray = buffer.copyOf(length)
}
